chess/position.go

```diff
@@ -0,0 +1,219 @@
+	p.Board.ClearPiece(fromPiece, m.From)
+
+	// Handle castling logic.
+	var castleRightsUsed CastleRights
+
+	switch fromPiece.Type {
+	case Rook:
+		// Moving a rook from its starting square relinquishes one castling right.
+		switch {
+		case fromPiece.Color == White && m.From == A1:
+			castleRightsUsed = WhiteQueenSide
+		case fromPiece.Color == White && m.From == H1:
+			castleRightsUsed = WhiteKingSide
+		case fromPiece.Color == Black && m.From == A8:
+			castleRightsUsed = BlackQueenSide
+		case fromPiece.Color == Black && m.From == H8:
+			castleRightsUsed = BlackKingSide
+		}
+	case King:
+		// King moves always relinquish both castling rights.
+		if fromPiece.Color == White {
+			castleRightsUsed = WhiteKingSide | WhiteQueenSide
+		} else {
+			castleRightsUsed = BlackKingSide | BlackQueenSide
+		}
+		// If this is a castle move, adjust the castling rook.
+		switch {
+		case m.From == E1 && m.To == G1:
+			p.Board.MovePieceToEmptySquare(Piece{White, Rook}, H1, F1)
+		case m.From == E1 && m.To == C1:
+			p.Board.MovePieceToEmptySquare(Piece{White, Rook}, A1, D1)
+		case m.From == E8 && m.To == G8:
+			p.Board.MovePieceToEmptySquare(Piece{Black, Rook}, H8, F8)
+		case m.From == E8 && m.To == C8:
+			p.Board.MovePieceToEmptySquare(Piece{Black, Rook}, A8, D8)
+		}
+	}
+
+	p.CastleRights.Clear(castleRightsUsed)
+
+	// Handle en passant capture.
+	if fromPiece.Type == Pawn && p.EnPassantRight.Valid && p.EnPassantRight.Square == m.To {
+		if fromPiece.Color == White {
+			p.Board.ClearPiece(Piece{Black, Pawn}, m.To.Below())
+		} else {
+			p.Board.ClearPiece(Piece{White, Pawn}, m.To.Above())
+		}
+	}
+
+	// Place the from piece on the to square.
+	if m.IsPromotion {
+		fromPiece.Type = m.Promotion
+	}
+	p.Board.SetPiece(fromPiece, m.To)
+
+	return
+}
+
+// A Board describes the placement of pieces.
+type Board struct {
+	Types  [6]Bitboard // Piece occupancies indexed by piece type.
+	Colors [2]Bitboard // Piece occupancies indexed by piece color.
+}
+
+// NewBoard returns a new board with all pieces in their starting positions.
+func NewBoard() Board {
+	return Board{
+		Types: [6]Bitboard{
+			0x00FF00000000FF00, // Pawns
+			0x4200000000000042, // Knights
+			0x2400000000000024, // Bishops
+			0x8100000000000081, // Rooks
+			0x0800000000000008, // Queens
+			0x1000000000000010, // Kings
+		},
+		Colors: [2]Bitboard{
+			0x000000000000FFFF, // White
+			0xFFFF000000000000, // Black
+		},
+	}
+}
+
+// Get returns the piece at the given square, if any.
+func (b *Board) Get(s Square) (Piece, bool) {
+	for t := Pawn; t <= King; t++ {
+		if b.Types[t].Get(s) {
+			isBlack := b.Colors[Black.Int()].Get(s)
+			return Piece{Color(isBlack), t}, true
+		}
+	}
+	return Piece{}, false
+}
+
+// SetPieceOnEmptySquare sets a piece on the given empty square.
+// When available, this is more performant than SetPiece.
+func (b *Board) SetPieceOnEmptySquare(p Piece, s Square) {
+	b.Types[p.Type].Set(s)
+	b.Colors[p.Color.Int()].Set(s)
+}
```

Comment:
I don't think we need two separate functions for setting pieces, moving pieces, etc.

One function is straightforward enough for now, and we should optimize more heavily later on

Reply:
Done! Simplified to just `Get`, `Set`, `Clear`, `Move`.